src/Native/System.Security.Cryptography.Native/pal_cms.cpp

```diff
@@ -103,3 +105,93 @@ extern "C" int CryptoNative_CmsGetRecipientStackFieldCount(CmsRecipientStack* re
 
     return sk_CMS_RecipientInfo_num(recipientStack);
 }
+
+extern "C" CMS_ContentInfo* CryptoNative_CmsEncryptInitialize(ASN1_OBJECT* algorithmOid, int32_t* status)
+{
+    if (algorithmOid == nullptr || status == nullptr)
+    {
+        *status = -1;
+        return nullptr;
+    }
+
+    const EVP_CIPHER* cipher = EVP_get_cipherbyobj(algorithmOid);
+
+    if (cipher == nullptr)
+    {
+        *status = 0;
+        return nullptr;
+    }
+
+    *status = 1;
+    return CMS_EnvelopedData_create(cipher);
+}
+
+static int CmsAddRecipients(CMS_ContentInfo* cms, X509Stack* certs, uint32_t flags)
+{
+    if (cms == nullptr || certs == nullptr)
+    {
+        return -1;
+    }
+
+    for (int i = 0; i < sk_X509_num(certs); i++) {
+        X509* recip = sk_X509_value(certs, i);
+        if (CMS_add1_recipient_cert(cms, recip, flags) == nullptr) {
+            CMSerr(CMS_F_CMS_ENCRYPT, CMS_R_RECIPIENT_ERROR);
+            return 0;
+        }
+    }
+
+    return 1;
+}
+
+extern "C" int CryptoNative_CmsAddSkidRecipients(CMS_ContentInfo* cms, X509Stack* skidCerts)
+{
+    return CmsAddRecipients(cms, skidCerts, CMS_PARTIAL | CMS_USE_KEYID);
+}
+
+extern "C" int CryptoNative_CmsAddIssuerAndSerialRecipients(CMS_ContentInfo* cms, X509Stack* issuerSerialCerts)
+{
+    return CmsAddRecipients(cms, issuerSerialCerts, CMS_PARTIAL);
+}
+
+extern "C" int CryptoNative_CmsAddOriginatorCerts(CMS_ContentInfo* cms, X509Stack* originatorCerts)
+{
+    if (cms == nullptr || originatorCerts == nullptr)
+    {
+        return -1;
+    }
+
+    for (int i = 0; i < sk_X509_num(originatorCerts); i++) {
+        X509* recip = sk_X509_value(originatorCerts, i);
+        if (CMS_add1_cert(cms, recip) == 0) {
+            return 0;
+        }
+    }
+
+    return 1;
+}
+
+extern "C" int CryptoNative_CmsEncryptFinalize(CMS_ContentInfo* cms, BIO* data)
+{
+    if (cms == nullptr || data == nullptr)
+    {
+        return -1;
+    }
+
+    if (CMS_set_detached(cms, 0) == 0)
```

Comment:
I have no idea what this does.  What does it do?  Is this the appropriate place?  Is it always right?  No matter what the answers, I'm sure that a comment would be useful to the next guy.